client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -161,26 +158,60 @@ public void shutdown() {
             }
         }
 
+        @Override
+        @Synchronized
+        public void refresh() {
+            if (started.get()) {
+                this.scheduledExecutor.execute(this::getControllers);
+            }
+        }
+
         // The controller discovery API invoker.
         private void getControllers() {
+            log.debug("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
```

Comment:
This is accessing `client` outside the scope of a lock, and the member is non-final. So this can result in `getControllerServerList` being invoked on an uninitialized object. Either make client final or guard it with a lock an annotate it with the lock that is guarding it.